Add known Kestrel connection error types (#1548)
open-telemetry/semantic-conventions: docs/dotnet/dotnet-kestrel-metrics.md

| Value  | Description | Stability |
|---|---|---|
| `aborted_by_app` | The HTTP/1.1 connection was aborted when app code aborted an HTTP request with `HttpContext.Abort()`. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `app_shutdown_timeout` | The connection was aborted during app shutdown. During shutdown, the server stops accepting new connections and HTTP requests, and it is given time for active requests to complete. If the app shutdown timeout is exceeded, all remaining connections are aborted. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `closed_critical_stream` | A critical control stream for an HTTP/3 connection was closed. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `connection_reset` | The connection was reset while there were active HTTP/2 or HTTP/3 streams on the connection. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Author: Issue created at https://github.com/dotnet/aspnetcore/issues/58885

> | `error_after_starting_response` | An error occurred after the response was started, causing an abort of the HTTP/1.1 connection or multiplexed request stream. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer: What does "multiplexed request stream" mean here? I would have suspected it's same thing as "HTTP/2 or HTTP/3 stream", but I don't see why we wouldn't just abort the individual stream rather than the entire connection in that case. Reading the code, it looks like `error_after_starting_response` connection end reason only gets set via the `Http1Connection.DisableKeepAlive` override.

```suggestion
| `error_after_starting_response` | An error such as an unhandled application exception or invalid request body occurred after the response was started, causing an abort of the HTTP/1.1 connection. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

| `error_reading_headers` | An error occurred when decoding HPACK headers in an HTTP/2 `HEADERS` frame. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `error_writing_headers` | An error occurred when encoding HPACK headers in an HTTP/2 `HEADERS` frame. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `flow_control_queue_size_exceeded` | The connection exceeded the outgoing flow control maximum queue size and was closed with `INTERNAL_ERROR`. This can be caused by an excessive number of HTTP/2 stream resets. For more information, see [Microsoft Security Advisory CVE-2023-44487:](https://github.com/dotnet/runtime/issues/93303). | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `flow_control_window_exceeded` | The client sent more data than was available in the flow-control window. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: ```suggestion
| `flow_control_window_exceeded` | The client sent more data than allowed by the current flow-control window. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

| `invalid_request_headers` | The HTTP request contains invalid headers. This error can occur in a number of scenarios: a header might not be allowed by the HTTP protocol, such as a pseudo-header in the `HEADERS` frame of an HTTP/2 request. A header could also have an invalid value, such as a non-integer `content-length`, or a header name or value might contain invalid characters. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `invalid_request_line` | The first line of an HTTP/1.1 request was invalid, potentially due to invalid content or exceeding the allowed limit. Configured by `KestrelServerLimits.MaxRequestLineSize`. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `invalid_settings` | The connection received an HTTP/2 or HTTP/3 `SETTINGS` frame with invalid settings. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `invalid_stream_id` | An HTTP/2 stream with an invalid stream ID was received. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `invalid_window_update_size` | The server received an HTTP/2 `WINDOW_UPDATE` frame that caused a flow-control window to exceed the maximum size. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: Or 0
https://github.com/dotnet/aspnetcore/blob/3fe94d130d0449aea927042817faeaff5f71b263/src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs#L1130-L1156

| `missing_stream_end` | The connection received an HTTP/2 `HEADERS` frame for trailers without a stream end flag. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `output_queue_size_exceeded` | The connection exceeded the output queue size and was closed with `INTERNAL_ERROR`. This can be caused by an excessive number of HTTP/2 stream resets. For more information, see [Microsoft Security Advisory CVE-2023-44487:](https://github.com/dotnet/runtime/issues/93303). | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `request_headers_timeout` | Request headers timed out while waiting for headers to be received after the request started. Configured by `KestrelServerLimits.RequestHeadersTimeout`. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `response_content_length_mismatch` | The HTTP response body sent data that didn't match the response's `content-length` header. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `server_timeout` | The connection timed out with the `IConnectionTimeoutFeature`. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: Technically correct but not useful. What timeouts signal this?

Author: It's from here: https://github.com/dotnet/aspnetcore/blob/b7c8cf7506338b9cd5c4960bd99b03d190c3c805/src/Servers/Kestrel/Core/src/Internal/Infrastructure/TimeoutControl.cs#L312-L324

As far as I can tell nothing in aspnetcore calls `IConnectionTimeoutFeature.SetTimeout`. The API is there for apps to use.

How do you suggest the description is improved?

Reviewer 3: Resolving this, but if there are any suggestions on how to improve things, please bring them on as an issue or PR.

| `stream_creation_error` | The HTTP/3 connection received a stream that it wouldn't accept. For example, the client created duplicate control streams. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `stream_reset_limit_exceeded` | The connection received an excessive number of HTTP/2 stream resets and was closed with `ENHANCE_YOUR_CALM`. For more information, see [Microsoft Security Advisory CVE-2023-44487:](https://github.com/dotnet/runtime/issues/93303). | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `stream_self_dependency` | The connection received an HTTP/2 frame that caused a frame to depend on itself. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 2: were stream dependencies ever implemented?

Author: No. But validation is still there.

| `tls_handshake_failed` | An error occurred during the TLS handshake for a connection. Only reported for HTTP/1.1 and HTTP/2 connections. The TLS handshake for HTTP/3 is internal to QUIC transport. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `tls_not_supported` | A TLS handshake was received by an endpoint that isn't configured to support TLS. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `unexpected_end_of_request_content` | The HTTP request body ended before the data specified by `content-length` was received. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer: I think we only close the connection for this reason for HTTP/1.1, and I think this gets raised for an incomplete chunked request body too.

```suggestion
| `unexpected_end_of_request_content` | The HTTP/1.1 request body ended before the data specified by the `content-length` header or chunked transfer encoding mechanism was received. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
```

| `unexpected_frame` | An unexpected HTTP/2 or HTTP/3 frame type was received. The frame type is either unknown, unsupported, or invalid for the current stream state. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |
| `unknown_stream` | An HTTP/2 frame was received on an unknown stream. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

> | `write_canceled` | The cancellation of a response body write aborted the HTTP/1.1 connection. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 3: Here's how we can do it now: https://github.com/lmolkova/semantic-conventions/commit/4c406b032daac776fb2a2407e6db97a0bf66ab6f

BTW, I think we have a bug in the current semconv - we list `error.type` on metrics that don't have it (e.g. `kestrel.upgraded_connections` or `kestrel.active_connections`.

> | `write_canceled` | The cancellation of a response body write aborted the HTTP/1.1 connection. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Author: Only `kestrel.tls_handshake.duration` and `kestrel.connection.duration` use `error.type`. Also, after this change, they have different values. The new known values only apply to `kestrel.connection.duration`.

> | `write_canceled` | The cancellation of a response body write aborted the HTTP/1.1 connection. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Reviewer 3: thanks for the explanation!

Then this - https://github.com/lmolkova/semantic-conventions/commit/dd87fbbb00c6db3db1414e5cd0f38c8d2bbb16a5 ?

> | `write_canceled` | The cancellation of a response body write aborted the HTTP/1.1 connection. | ![Experimental](https://img.shields.io/badge/-experimental-blue) |

Author: Yes. I've updated the PR with your changes, and applied some other PR feedback. However there are a bunch of warnings that yaml lines are too long. What is the best way to fix that?

<!-- END AUTOGENERATED TEXT -->
<!-- endsemconv -->

> [DocumentStatus]: https://opentelemetry.io/docs/specs/otel/document-status

Author: ```suggestion
[DocumentStatus]: https://opentelemetry.io/docs/specs/otel/document-status

```
